inst/tests/tests.Rraw

```diff
@@ -5951,15 +5951,15 @@ setindex(DT,b)
 test(1420, indices(DT), "b")
 setnames(DT,"b","foo")
 test(1421, indices(DT), "foo")
-test(1422, DT[foo==9, verbose=TRUE], DT[2], output="Using existing index 'foo'")
+test(1422, DT[foo==9, verbose=TRUE], DT[2], output="Optimized subsetting with index '__foo'")
 setindex(DT,a,foo)
 test(1423, indices(DT), c("foo","a__foo"))   # tests as well that order of attributes is retained although we don't use that property currently.
 test(1424, indices(setnames(DT,"foo","bar")), c("bar","a__bar"))
 test(1425, indices(setnames(DT,"a","baz")), c("bar","baz__bar"))
-test(1426, DT[baz==4L, verbose=TRUE], output="Creating new index 'baz'")
-test(1427, indices(DT), c("bar","baz__bar","baz"))
-test(1428, DT[bar==9L, verbose=TRUE], output="Using existing index 'bar'")
-test(1429, indices(setnames(DT,"bar","a")), c("baz", "a", "baz__a"))
+test(1426, DT[baz==4L, verbose=TRUE], output="Optimized subsetting with index '__baz__bar'")
```

Comment:
why this test doesn't create new index anymore but use existing one? it is potentially breaking change

Reply:
As I argue in the first comment of this PR, I think there is no chance that this is a breaking change:
>Internally, indices are now used, even if not all index columns appear in the query.
>Previously, this was prevented out of fear for spurious reordering. This, however, can't happen since the order is restored brute force after bmerge (unit tested in the PR).

The unit test that I inserted to guarantee no reordering is this one:
```
## test no reordering in groups
 DT <- data.table(x = c(1,1,1,2), y = c(3,2,1,1))
 setindex(DT, x, y)
 test(1437.3, DT[x==1], setindex(data.table(x = c(1,1,1), y = c(3,2,1)), x,y))
```
What do you think?

Reply:
Adding new tests is fine but when changing old one you should have a reason. I don't understand why 'creating index' has change to 'using index' in 1426. Maybe it is copy-paste mistakes which pass the tests because you create and use index in that test.

Reply:
Sorry, I wasn't clear enough.
In the master, the index  `__baz` is created in 1426. In the speedySubset, however, no index is created, as the existing index `__baz__bar` is used:
```
## on branch speedySubset, repeating test 1426
DT = data.table(a=1:5,b=10:6)
setindex(DT,b)
setnames(DT,"b","foo")
setindex(DT,a,foo)
setnames(DT,"foo","bar")
setnames(DT,"a","baz")
DT[baz==4L, verbose=TRUE]
# Optimized subsetting with index 'baz__bar'
# Starting bmerge ...done in 0.001 secs
```
As argued above, I believe that this doesn't cause any problems since the original order of the data is restored brute force by the line:
`if (length(o$xo)) i = fsort(o$xo[i], internal=TRUE) else i = fsort(i, internal=TRUE) # fix for #1495`
If you have the feeling that this change is too dangerous because of unknown side-effects that are not captured by any of the unit-tests, I will revert the change. It was just an optimization during cleanup and not the main purpose of the PR.

Reply:
@MarkusBonsch It is very good that we don't need to create new index and can reuse part of the existing index. I was not aware of this improvement in PR. I did not look at the ordering issue and cannot say anything about it now.
What if user will subset many times by `baz` field? Then it might be better to create this single column index and not to maintain original order by brute force?
To handle that well every multi-column index should include order of its parent subsets of columns within the index. This wouldn't cost much to create (as it has to computed for the desired index anyway) but it will take much more memory.

Reply:
@jangorecki You are right that brute force reordering is costly, as has been shown here: #2366 and also in the benchmarks of this PR in #2472
However, it is an existing part of the master and not introduced in this PR.
No matter, which index we use, `baz` or `baz__bar`, this brute force reordering is necessary to ensure the original order of the DT. Which index is faster in terms of reordering depends on the original order of the DT. In my opinion, there is no clear indication that a single-column index will be substantially more ordered than a multi-column index in general. If a user has to do many subsets on `baz`, the fastest option is a key since this will avoid reordering costs almost completely.